cpp/src/shuffler/chunk.cpp

```diff
@@ -14,6 +14,37 @@
 
 namespace rapidsmpf::shuffler::detail {
 
+
+/**
+ * @brief Try to make a memory reservation.
+ *
+ * @param br The buffer resource.
+ * @param size The size of the buffer to allocate.
+ * @param preferred_mem_type The preferred memory type to allocate the buffer from.
+ * @return A memory reservation.
+ * @throw RAPIDSMPF_FAIL if no memory reservation was made.
+ */
+MemoryReservation try_reserve_or_fail(
+    BufferResource* br,
+    size_t size,
+    std::optional<MemoryType> const& preferred_mem_type = std::nullopt
+) {
+    if (preferred_mem_type) {
+        auto [res, overbooking] = br->reserve(*preferred_mem_type, size, false);
+        RAPIDSMPF_EXPECTS(res.size() == size, "failed to reserve memory");
+        return std::move(res);
+    }
+
+    // try to allocate data buffer from memory types in order [DEVICE, HOST]
+    for (auto mem_type : MEMORY_TYPES) {
+        auto [res, overbooking] = br->reserve(mem_type, size, false);
+        if (res.size() == size) {
+            return std::move(res);
+        }
+    }
+    RAPIDSMPF_FAIL("failed to reserve memory");
```

Comment:
Throw something other than `std::logic_error`?
```suggestion
    RAPIDSMPF_FAIL("failed to reserve memory", std::runtime_error);
```